flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.datagen.source;
+
+import org.apache.flink.annotation.Experimental;
+import org.apache.flink.api.connector.source.ReaderOutput;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.api.connector.source.SourceReaderContext;
+import org.apache.flink.api.connector.source.lib.util.IteratorSourceReaderBase;
+import org.apache.flink.api.connector.source.lib.util.IteratorSourceSplit;
+import org.apache.flink.core.io.InputStatus;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.util.Iterator;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BooleanSupplier;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A {@link SourceReader} that synchronizes emission of N elements on the arrival of the checkpoint
+ * barriers It 1) emits a list of elements without checkpoints in-between, 2) then waits for two
+ * checkpoints to complete, 3) then re-emits the same elements before 4) waiting for another two
+ * checkpoints and 5) exiting.
+ *
+ * <p>This lockstep execution is possible because {@code pollNext} and {@code snapshotState} are
+ * executed in the same thread and the fact that {@code pollNext} can emit N elements at once. This
+ * reader is meant to be used solely for testing purposes as the substitution for the {@code
+ * FiniteTestSource} which implements the deprecated {@code SourceFunction} API.
+ */
+@Experimental
+public class SourceReaderWithCheckpointsLatch<
+                E, O, IterT extends Iterator<E>, SplitT extends IteratorSourceSplit<E, IterT>>
+        extends IteratorSourceReaderBase<E, O, IterT, SplitT> {
+
+    private final GeneratorFunction<E, O> generatorFunction;
+
+    private BooleanSupplier couldExit;
+    private int snapshotsCompleted;
+    private int snapshotsToWaitFor = Integer.MAX_VALUE;
+    private boolean done;
+
+    public SourceReaderWithCheckpointsLatch(
+            SourceReaderContext context,
+            GeneratorFunction<E, O> generatorFunction,
+            @Nullable BooleanSupplier couldExit) {
```

Comment:
```suggestion
            @Nullable BooleanSupplier safeToExit) {
```

```suggestion
            @Nullable BooleanSupplier canExit) {
```

```suggestion
            @Nullable BooleanSupplier shouldExit) {
```


These make more sense to me.

Reply:
Good idea, it is based on the existing naming in FiniteTestSource
https://github.com/apache/flink/blob/0234027c55ce743e3340391c77348f100f712e90/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/FiniteTestSource.java#L77
But changing it makes sense.